### DISCUSSION TITLE
How to start a service for 3 different runners?

### DISCUSSION THREAD

**Discussion by BoboTiG** [community/community#25124] [Actions]
<p>Hello,</p>
<p>I did not find suitable answers on the internet, so here I am :slight_smile:</p>
<p>My goal is simple to explain:</p>
<ul>
<li>I am testing a multiplaform application (GNU/Linux, macOS and Windows).</li>
<li>That application needs to be connect to a server.</li>
</ul>
<p>So I need to start the server first, then the 3 other OSes.</p>
<p>Where I struggle is how to start the server first? The server is a docker image, so it can be run from a <code>services</code>, but only on GNU/Linux (and maybe Windows, but not tested; and this is not possible on macOS) …</p>
<hr>
<p>This is what I’ve tested (simplified):</p>
<pre><code>jobs:
  # start_server:
  # runs-on: ubuntu-latest
  # services:
  # image: nuxeo/nuxeo:master
  # ports:
  # - 8080:8080
  # # env:
  # # ADDITIONAL_PACKAGES: nuxeo-drive nuxeo-platform-importer
  # # NUXEO_CLID: ${{ secrets.NUXEO_CLID }}

  ft_linux:
    needs: start_server
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
    - uses: actions/setup-python@v1
    - name: Functional tests
      run: tox -e ft

  ft_macos:
    needs: start_server
    runs-on: macos-latest
    steps:
    - uses: actions/checkout@v2
    - uses: actions/setup-python@v1
    - name: Functional tests
      run: tox -e ft

  ft_windows:
    needs: start_server
    runs-on: windows-latest
    steps:
    - uses: actions/checkout@v2
    - uses: actions/setup-python@v1
    - name: Functional tests
      run: tox -e ft
</code></pre>
<p>This does not work because the docker container will never stop (it is a webserver, this is expected), and so the <code>start_server</code> action will never give the hand to next actions.</p>
<p>Thanks a lot for your inputs.</p>

**Comment by BoboTiG** [comment]
<p>I forgot to say: this does not work because the docker container will never stop (it is a webserver, this is expected), and so the <code>start_server</code> action will never give the hand to next actions.</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/bobotig">@bobotig</a> ,</p>
<p>Here few points need you pay attention to when you use GitHub-hosted runners to run the workflows:<br>
<strong>1)</strong> Each job in a workflow executes in a fresh instance of the virtual machine.<br>
<strong>2)</strong> The virtual machine will be discarded after the job completed, all the configurations in the virtual machine also will be discarded together.<br>
<strong>3)</strong> Service containers are only available to Linux runners.</p>
<p>According to above points, I’m afraid it’s hard to meet your request.<br>
Maybe you can try using a self-hosted Linux runner to run the service container job, because the self-hosed runner machine will not be discarded after the job completed. Maybe this can keep the container active on the machine instead of being discarded.</p>
<p>In addition, you need to understand that each job in the workflow must have at least one step.</p>

**Comment by BoboTiG** [comment]
<p>Thanks <a class="mention" href="/brightran">@brightran</a>,</p>
<p>I now have the confirmation about the feasability of the workflow, not an easy one :slight_smile:</p>
<p>It is OK for the job’s step, I just removed most of the lines to help keep the example easy to read.</p>
<p>Regards,</p>
